core/src/loader.rs

```diff
@@ -1916,6 +1956,76 @@ impl<'gc> Loader<'gc> {
         Ok(())
     }
 
+    /// This makes the MovieClip enter the initial loading state in which some
+    /// attributes have certain initial loading values to signal that the file is
+    /// currently being loaded and neither an error has occurred nor the first frame
+    /// has been successfully loaded yet.
+    fn load_initial_loading_swf(
+        mc: &mut MovieClip<'gc>,
+        uc: &mut UpdateContext<'_, 'gc>,
+        request_url: &str,
+        resolved_url: Result<Url, ParseError>,
+    ) {
+        match resolved_url {
+            Err(_) => {
+                Loader::load_error_swf(mc, uc, request_url.to_string());
+            }
+            Ok(url) => {
+                // If the loaded SWF is a local file, the initial loading state equals the error state.
+                if url.scheme() == "file" {
+                    Loader::load_error_swf(mc, uc, url.to_string());
+                } else {
+                    // Replacing the movie sets total_frames and and frames_loaded correctly.
+                    // The movie just needs to be the default empty movie with the correct URL.
+                    // In this loading state, the URL is the URL of the parent movie / doesn't change.
+
+                    let current_movie = mc.movie();
+                    let current_version = current_movie.version();
+                    let current_url = current_movie.url();
+                    let mut initial_loading_movie = SwfMovie::empty(current_version);
+                    initial_loading_movie.set_url(current_url.to_string());
+
+                    mc.replace_with_movie(uc, Some(Arc::new(initial_loading_movie)), None);
+
+                    // Maybe this (keeping the current URL) should be the default behaviour
+                    // of replace_with_movie?
+                    // TODO: See where it gets invoked without a movie as well and what the
+                    // correct URL result is in these cases.
+                }
+            }
+        }
+    }
+
+    /// This makes the MovieClip enter the error state in which some attributes have
+    /// certain error values to signal that no valid file could be loaded.
+    /// An error state movie stub which provides the correct values is created and
+    /// loaded.
+    ///
+    /// This happens if no file could be loaded or if the loaded content is no valid
+    /// supported content.
+    ///
+    /// swf_url is always the final URL obtained after any redirects.
+    fn load_error_swf(
+        mc: &mut MovieClip<'gc>,
+        uc: &mut UpdateContext<'_, 'gc>,
+        mut swf_url: String,
+    ) {
+        // If a local URL is fetched using the flash plugin, the _url property
+        // won't be changed => It keeps being the parent SWF URL.
+        if cfg!(target_family = "wasm") {
```

Comment:
How would you suggest that? I could imagine a `NavigatorBackend` method like `isOnWeb` which returns true in the `WebNavigatorBackend` and false in the others. If the wasm check isn't clean to test whether we're in a browser, this might be a better idea (although a cfg like this is used in two other places in Ruffle; it might make sense to replace them as well).

The test itself should be done in this method; otherwise it would need to be implemented in different places for both the initial loading and the error state since this applies to both and the method is called for both (if the file is local).

Reply:
Checking the compile architecture to determine the build type is something that we do in many places.

Reply:
We might want to have an explicit feature for at least some of the things that currently rely on `target_family='wasm'`. Again, entirely out of scope. For this particular PR alone I was more confused as to why `NavigatorBackend`'s fancy new `parse_url` method wasn't being used to change `file://` URLs.

Reply:
The new `resolve_url` method is just used to actually resolve the url string and do the necessary preprocessing, which depends on the `NavigatorBackend` implementation. It is used by both the `navigate_to_url` (which opens a URL link) and the `fetch` (which loads a movie) methods.
These methods need to know the correct resolved URL. E.g. if `navigate_to_url` is called with a local URL string, it needs to know that it's local, throws an error and doesn't open it.
What this code does, is setting the `SwfMovie::url` variable. And if ActionScript code tries to load a local file on web, Flash Player makes the MovieClip go into the error state but doesn't change the url, FP "lies about it" and it just stays the same.

But this just works if `resolve_url` returns the correct `Url` and this condition is specifically tested afterwards. If `resolve_url` itself would return this in the situation, `fetch` and `navigate_to_url` would think it's a valid url (the url of the parent SWF itself) and load / navigate there. So yeah, they need to know it's a local file / invalid and `resolve_url` needs to actually just resolve the URL correctly, but because FP lies in the error state, when setting this state, this needs to be deliberately tested and taken into account to set the state as FP does.

Reply:
Gotcha.

Reply:
The fetch API doesn't allow access to `file` URLs, so all of this is really inconsequential/pointless, right?

Reply:
@n0samu What exactly do you mean? The code highlighted in this review sets the `url` property of the `MovieClip, which is accessible in ActionScript. If this isn't done, the `url` property of a `MovieClip` is wrong under these specific circumstances.
In general, the `NavigationBackend` refactor allows for better code quality and improves the error messages. Previously, you could also not load local files on web, but a generic "Got JS error" got returned, while now that case is directly handled when resolving the URL and a precise error message is returned.

Reply:
Oh okay, thanks for the explanation, don't mind me

Reply:
Dw; it's a good question
And you're welcome :)